src/workqueue/workqueue.c

```diff
@@ -0,0 +1,833 @@
+ * the scheduling management subsystem.
+ * TODO: This may be moved "down" to user-level scheduler
+ *       if it is not easy to maintain from the kext.
+ */
+static volatile int32_t _dispatch_workq_runnable_workers;
+
+/* The desired minimum number of runnable worker threads */
+static int32_t _dispatch_workq_min_runnable_workers;
+
+/* The desired maximum number of runnable worker threads */
+static int32_t _dispatch_workq_max_runnable_workers;
+
+/* Limit on the total number of worker threads that can be created. */
+static int32_t _dispatch_workq_max_spawned_workers;
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+/* Are we using user-level or kext based management? */
+static bool _dispatch_workq_kext_active;
+#endif
+
+/* Called on startup of the workqueue system */
+static void _dispatch_workq_management_init(void);
+
+/* Called by worker thread when it starts executing */
+static int _dispatch_workq_worker_register(void);
+
+/* Called by worker thread just before it exits */
+static void _dispatch_workq_worker_unregister(void);
+
+/* Called when new work is added to the system */
+static void _dispatch_workq_work_added(void);
+
+/*
+ * Called by a worker thread when either it has found no work to do
+ * or it has been command to wait by the management subsystem.
+ * This call is expected to block until the management subsystem
+ * thinks there may be work for the thread to do.
+ */
+static void _dispatch_workq_worker_wait(bool was_ordered_to_wait);
+
+/*
+ * Called by the management system when it wants to inject a
+ * workitem that will case a worker thread to wait
+ */
+static void _dispatch_workq_worker_should_wait(void);
+
+/*
+ * Called by the management system when it wants to inject a
+ * workitem that will case a worker thread to terminate
+ */
+static void _dispatch_workq_worker_should_exit(void);
+
+/*
+ * Called by the management system when it wants to
+ * create an additional worker thread.
+ */
+static void _dispatch_workq_worker_should_start(void);
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+/*
+ * outlined helpers for above functions to temporarily
+ * isolate kext implementation code to separate file.
+ */
+int _dispatch_workq_management_init_kext(void);
+void _dispatch_workq_worker_register_kext(void);
+void _dispatch_workq_worker_unregister_kext(void);
+void _dispatch_workq_work_added_kext(void);
+void _dispatch_workq_worker_wait_kext(void);
+void* _dispatch_workq_thread_monitor_main_kext(void *context DISPATCH_UNUSED);
+#endif // DISPATCH_ENABLE_PWQ_KEXT
+
+#pragma mark Workqueue internal data structures and functions
+
+typedef struct dispatch_workq_item_s {
+	STAILQ_ENTRY(dispatch_workq_item_s) item_entry;
+	void (*func)(void *);
+	void *func_arg;
+} dispatch_workq_item_s, *dispatch_workq_item_t;
+
+typedef struct dispatch_workq_workqueue_s {
+	STAILQ_HEAD(,dispatch_workq_item_s) item_listhead;
+	dispatch_unfair_lock_s lock;
+	int priority;
+	bool overcommit;
+} dispatch_workq_workqueue_s, *dispatch_workq_workqueue_t;
+
```

Comment:
went with dispatch_workq_context_t for now (missed the edited comment).  

I don't expect to actually need this type in the long run.  For the initial merge, I was staying pretty close to the libpwq structure which has a redundant layer of queuing.  I was planning on eliminating that redundancy in a second round of changes.  We can talk on Monday about whether this plan makes sense or if we should jump to something closer to the final structure in one big step.